test/alternator/test_returnvalues.py

```diff
@@ -83,6 +83,47 @@ def test_put_item_returnvalues(test_table_s):
     with pytest.raises(ClientError, match='ValidationException'):
         test_table_s.put_item(Item={'p': p, 'a': 'hello'}, ReturnValues='none')
 
+def skip_if_returnvalues_on_condition_check_not_supported():
+    import botocore
+    from packaging.version import Version
+    # This release added support for ReturnValuesOnConditionCheckFailure
+    if (Version(botocore.__version__) < Version('1.29.164')):
+        pytest.skip("Botocore version 1.29.164 or above required to run this test")
+
+# Testing ReturnValuesOnConditionCheckFailure feature which returns values only
+# on failed condition expression.
+def test_put_item_returnvalues_on_condition_check_failure(test_table_s):
+    skip_if_returnvalues_on_condition_check_not_supported()
+    p = random_string()
+    test_table_s.put_item(Item={'p': p, 's': 'dog'})
+    # Succesful conditional put doesn't return values.
+    ret=test_table_s.put_item(Item={'p': p, 's': 'cat'},
+        ReturnValuesOnConditionCheckFailure='ALL_OLD',
+        ConditionExpression='s = :v1',
+        ExpressionAttributeValues={':v1' : 'dog'})
+    assert not 'Attributes' in ret # Ordinary field not present.
+    assert not 'Item' in ret # Field used for error responses also not present.
+    # Failed conditional put returns old values.
+    with pytest.raises(test_table_s.meta.client.exceptions.ConditionalCheckFailedException) as err:
+        test_table_s.put_item(Item={'p': p, 's': 'cow'},
+            ReturnValuesOnConditionCheckFailure='ALL_OLD',
+            ConditionExpression='s = :v1',
+            ExpressionAttributeValues={':v1' : 'dog'})
+    assert err.value.response['Item'] == {'p': {'S': p}, 's': {'S': 'cat'}}
+    # Failed conditional put without returning old values.
+    with pytest.raises(test_table_s.meta.client.exceptions.ConditionalCheckFailedException) as err:
+        test_table_s.put_item(Item={'p': p, 's': 'cow'},
+            ReturnValuesOnConditionCheckFailure='NONE',
+            ConditionExpression='s = :v1',
+            ExpressionAttributeValues={':v1' : 'dog'})
+    assert not 'Item' in err.value.response
+    # Invalid ReturnValuesOnConditionCheckFailure value returns error.
+    with pytest.raises(ClientError, match='ValidationException'):
+        test_table_s.put_item(Item={'p': p, 's': 'cow'},
+            ReturnValuesOnConditionCheckFailure='DUMMY',
+            ConditionExpression='s = :v1',
+            ExpressionAttributeValues={':v1' : 'dog'})
```

Comment:
Thanks for adding all these tests.
I have another idea for a test: What if ReturnValues (not ReturnValuesOnConditionCheckFailure!) is set, and the condition failed - will it set the Item on the error or not?
If the answer is no, does combining ReturnValues and ReturnValuesOnConditionCheckFailure work? (it should return the item both in the successful case and error case).

Reply:
this gets exponential I am afraid, since those features use different fields both for setup and output we can assume they are orthogonal, otherwise we won't ever stop adding new test cases

Reply:
One of the idea of making Alternator tests quick to run (right now we have 739 test functions taking 30 seconds sequentially), and quick to write, is that we don't need to be afraid to add new test cases.
I agree that we don't want to explore all combinations of different features - e.g., ReturnValuesOnConditionCheckFailure combined with a materialized view isn't particularly interesting. But when you as a developer know that two features are risky together because they touch the same code paths (e.g., ReturnValues and ReturnValuesOnConditionCheckFailure both mess with the same variable) or when a test checks some "if" case but doesn't check its "else" case, it can be useful to add more tests.
I don't insist on these additional tests - I agree that improving the test coverage is a never-ending effort and we can always add more later. But if you're looking for ideas on what you didn't test and might test - I wanted to suggest ideas.